metagraph/src/cli/query.cpp

```diff
@@ -263,29 +263,27 @@ void call_hull_sequences(const DeBruijnGraph &full_dbg,
  *             from the full annotation matrix
  *
  * @param[in]  full_annotation  The full annotation matrix.
- * @param[in]  index_in_full    Indexes of rows in the full annotation matrix.
- *                              index_in_full[i] = -1 means that the i-th row in
- *                              the submatrix is empty.
+ * @param[in]  num_rows         The number of rows in the target submatrix.
+ * @param[in]  full_to_small    The mapping between the rows in the full matrix
+ *                              and its submatrix.
  * @param[in]  num_threads      The number of threads used.
  *
  * @return     Annotation submatrix in the UniqueRowAnnotator representation
  */
 std::unique_ptr<annot::UniqueRowAnnotator>
 slice_annotation(const AnnotatedDBG::Annotator &full_annotation,
-                 const std::vector<uint64_t> &index_in_full,
+                 uint64_t num_rows,
+                 std::vector<std::pair<uint64_t, uint64_t>>&& full_to_small,
                  size_t num_threads) {
-    const uint64_t npos = -1;
-
     if (auto *rb = dynamic_cast<const RainbowMatrix *>(&full_annotation.get_matrix())) {
         // shortcut construction for Rainbow<> annotation
-        std::vector<uint64_t> row_indexes;
-        row_indexes.reserve(index_in_full.size());
-        for (uint64_t i : index_in_full) {
-            if (i != npos) {
-                row_indexes.push_back(i);
-            } else {
-                row_indexes.push_back(0);
-            }
+        std::vector<bool> empty_rows(num_rows, true);
```

Comment:
Why not just initialize `row_indices` to `full_annotation.num_objects()` and use that to indicate empty rows?

Reply:
`get_rows` requires that all indexes in `row_indices` are valid.